src/protocol/ipa/mod.rs

```diff
@@ -0,0 +1,334 @@
+        match self {
+            Self::MatchKeyShares => "match_key_shares",
+            Self::TriggerBit => "is_trigger_bit",
+            Self::BreakdownKey => "breakdown_key",
+            Self::TriggerValue => "trigger_value",
+        }
+    }
+}
+
+pub struct IPAInputRow<F: Field> {
+    mk_shares: XorReplicated,
+    is_trigger_bit: Replicated<F>,
+    breakdown_key: Replicated<F>,
+    trigger_value: Replicated<F>,
+}
+
+struct IPAModulusConvertedInputRow<F: Field> {
+    mk_shares: Vec<Replicated<F>>,
+    is_trigger_bit: Replicated<F>,
+    breakdown_key: Replicated<F>,
+    trigger_value: Replicated<F>,
+}
+
+#[async_trait]
+impl<F: Field> Resharable<F> for IPAModulusConvertedInputRow<F>
+where
+    F: Sized,
+{
+    type Share = Replicated<F>;
+
+    async fn reshare<C>(&self, ctx: C, record_id: RecordId, to_helper: Role) -> Result<Self, Error>
+    where
+        C: Context<F, Share = <Self as Resharable<F>>::Share> + Send,
+    {
+        let f_mk_shares = self.mk_shares.reshare(
+            ctx.narrow(&IPAInputRowResharableStep::MatchKeyShares),
+            record_id,
+            to_helper,
+        );
+        let f_is_trigger_bit = ctx.narrow(&IPAInputRowResharableStep::TriggerBit).reshare(
+            &self.is_trigger_bit,
+            record_id,
+            to_helper,
+        );
+        let f_breakdown_key = ctx
+            .narrow(&IPAInputRowResharableStep::BreakdownKey)
+            .reshare(&self.breakdown_key, record_id, to_helper);
+        let f_trigger_value = ctx
+            .narrow(&IPAInputRowResharableStep::TriggerValue)
+            .reshare(&self.trigger_value, record_id, to_helper);
+
+        let (mk_shares, mut outputs) = try_join(
+            f_mk_shares,
+            try_join_all([f_is_trigger_bit, f_breakdown_key, f_trigger_value]),
+        )
+        .await?;
+
+        Ok(IPAModulusConvertedInputRow {
+            mk_shares,
+            is_trigger_bit: outputs.remove(0),
+            breakdown_key: outputs.remove(0),
+            trigger_value: outputs.remove(0),
+        })
+    }
+}
+
+/// # Errors
+/// Propagates errors from multiplications
+#[allow(dead_code)]
+pub async fn ipa<F>(
+    ctx: SemiHonestContext<'_, F>,
+    input_rows: &[IPAInputRow<F>],
+    num_bits: u32,
+    per_user_credit_cap: u32,
+) -> Result<Vec<AggregateCreditOutputRow<F>>, Error>
+where
+    F: Field,
+{
+    let mk_shares = input_rows.iter().map(|x| x.mk_shares).collect::<Vec<_>>();
+    let local_lists = convert_all_bits_local(ctx.role(), &mk_shares, num_bits);
+    let converted_shares = convert_all_bits(
+        &ctx.narrow(&Step::ModulusConversionForMatchKeys),
+        &local_lists,
+    )
+    .await
+    .unwrap();
+    let sort_permutation = generate_permutation_and_reveal_shuffled(
+        ctx.narrow(&Step::GenSortPermutationFromMatchKeys),
+        &converted_shares,
+        num_bits,
+    )
+    .await
+    .unwrap();
+
+    let converted_shares = transpose(&converted_shares);
```

Comment:
Can we flag this as a potential problem?  The use of transposition means a lot of allocations and a break from streaming that might be a performance problem later.

I think that ultimately what we'll want to do is provide the sort functions with entire input rows, with a means of it keying into that input row with a bit index (or radix N index, I guess).

Reply:
You're absolutely right. This is totally sub-optimal. I've filed #330 to track this issue.